snowfall/training/hmm_topo.py

```diff
@@ -0,0 +1,48 @@
+import k2
+from typing import List
+
+
+def build_hmm_topo_2state(tokens: List[int]) -> k2.Fsa:
+    """
+    Build a 2-state HMM topology used in Kaldi's chain models.
+    The first HMM state is entered only once for each token instance,
+    and the second HMM state is self-looped and optional.
+
+    Args:
+        tokens:
+            A list of token int IDs, e.g., phones, characters, etc.
+            The IDs for the first HMM state will be the same as token IDs;
+            The IDs for the second HMM state are: ``token_id + len(tokens)``
+    Returns:
+        An FST that converts a sequence of HMM state IDs to a sequence of token IDs.
+    """
+    followup_tokens = range(len(tokens), len(tokens) * 2)
```

Comment:
should it be
```python
followup_tokens = range(len(tokens) + 1, len(tokens) * 2 + 1)
```
as token id starts from 1?

Reply:
Yes, you're right. In the general case, to avoid surprises, I think that should be `len(tokens) + min_token_id`.